From c8a9e221a2515192feb28d4db8bcad620fbbd2f8 Mon Sep 17 00:00:00 2001
From: cjybyjk <cjybyjk@zjnu.edu.cn>
Date: Wed, 7 Jul 2021 22:07:55 +0800
Subject: [PATCH 091/103] Screenrecord: Export RecordingService for External
 Usage

[nullxception: simplify start handling and adapt it to 12L]
[neobuddy89: Adapted to 13]

Change-Id: I58be35cce6bf5a9706fce89327df5e215ca3f9bc
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 packages/SystemUI/AndroidManifest.xml         | 11 ++-
 .../screenrecord/IRecordingCallback.aidl      |  6 ++
 .../screenrecord/IRemoteRecording.aidl        | 12 +++
 .../screenrecord/RecordingService.java        | 81 ++++++++++++++++++-
 4 files changed, 108 insertions(+), 2 deletions(-)
 create mode 100644 packages/SystemUI/src/com/android/systemui/screenrecord/IRecordingCallback.aidl
 create mode 100644 packages/SystemUI/src/com/android/systemui/screenrecord/IRemoteRecording.aidl

diff --git a/packages/SystemUI/AndroidManifest.xml b/packages/SystemUI/AndroidManifest.xml
index e7cc18748d91..346758e365f4 100644
--- a/packages/SystemUI/AndroidManifest.xml
+++ b/packages/SystemUI/AndroidManifest.xml
@@ -207,6 +207,9 @@
     <permission android:name="android.permission.CUSTOMIZE_SYSTEM_UI"
         android:protectionLevel="signature|privileged" />
 
+    <permission android:name="com.android.systemui.permission.SCREEN_RECORDING"
+            android:protectionLevel="signature" />
+
     <!-- Adding Quick Settings tiles -->
     <uses-permission android:name="android.permission.BIND_QUICK_SETTINGS_TILE" />
 
@@ -501,7 +504,13 @@
             android:permission="android.permission.LAUNCH_CAPTURE_CONTENT_ACTIVITY_FOR_NOTE" />
 
         <service android:name=".screenrecord.RecordingService"
-                 android:foregroundServiceType="systemExempted"/>
+            android:foregroundServiceType="systemExempted"
+            android:exported="true"
+            android:permission="com.android.systemui.permission.SCREEN_RECORDING">
+            <intent-filter>
+                <action android:name="com.android.systemui.screenrecord.RecordingService" />
+            </intent-filter>
+        </service>
 
         <service android:name=".recordissue.IssueRecordingService"
                  android:foregroundServiceType="systemExempted"/>
diff --git a/packages/SystemUI/src/com/android/systemui/screenrecord/IRecordingCallback.aidl b/packages/SystemUI/src/com/android/systemui/screenrecord/IRecordingCallback.aidl
new file mode 100644
index 000000000000..ee4a46393989
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/screenrecord/IRecordingCallback.aidl
@@ -0,0 +1,6 @@
+package com.android.systemui.screenrecord;
+
+interface IRecordingCallback {
+    void onRecordingStart();
+    void onRecordingEnd();
+}
diff --git a/packages/SystemUI/src/com/android/systemui/screenrecord/IRemoteRecording.aidl b/packages/SystemUI/src/com/android/systemui/screenrecord/IRemoteRecording.aidl
new file mode 100644
index 000000000000..13cbf8dfb39b
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/screenrecord/IRemoteRecording.aidl
@@ -0,0 +1,12 @@
+package com.android.systemui.screenrecord;
+
+import com.android.systemui.screenrecord.IRecordingCallback;
+
+interface IRemoteRecording {
+    void startRecording();
+    void stopRecording();
+    boolean isRecording();
+    boolean isStarting();
+    void addRecordingCallback(in IRecordingCallback callback);
+    void removeRecordingCallback(in IRecordingCallback callback);
+}
diff --git a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
index 7c1d35452736..4be6c2fe2d9f 100644
--- a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
+++ b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
@@ -24,6 +24,7 @@ import android.app.PendingIntent;
 import android.app.Service;
 import android.content.ContentResolver;
 import android.content.Context;
+import android.content.ComponentName;
 import android.content.Intent;
 import android.graphics.PixelFormat;
 import android.graphics.drawable.Icon;
@@ -61,6 +62,7 @@ import com.android.systemui.settings.UserContextProvider;
 import com.android.systemui.statusbar.phone.KeyguardDismissUtil;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.concurrent.Executor;
 
 import javax.inject.Inject;
@@ -95,11 +97,13 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
     protected static final String ACTION_STOP = "com.android.systemui.screenrecord.STOP";
     protected static final String ACTION_STOP_NOTIF =
             "com.android.systemui.screenrecord.STOP_FROM_NOTIF";
+    private static final String ACTION_SHOW_DIALOG = "com.android.systemui.screenrecord.SHOW_DIALOG";
     protected static final String ACTION_SHARE = "com.android.systemui.screenrecord.SHARE";
     protected static final String ACTION_DELETE = "com.android.systemui.screenrecord.DELETE";
     private static final String PERMISSION_SELF = "com.android.systemui.permission.SELF";
     private static final String PREF_DOT_RIGHT = "screenrecord_dot_right";
 
+    private final RecordingServiceBinder mBinder;
     private final RecordingController mController;
     protected final KeyguardDismissUtil mKeyguardDismissUtil;
     private final Handler mMainHandler;
@@ -137,6 +141,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
         mWindowManager = (WindowManager) userContextTracker.getUserContext()
                 .getSystemService(Context.WINDOW_SERVICE);
         mKeyguardDismissUtil = keyguardDismissUtil;
+        mBinder = new RecordingServiceBinder();
     }
 
     /**
@@ -291,18 +296,30 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 // Close quick shade
                 maybeCloseSystemDialogs();
                 break;
+            case ACTION_SHOW_DIALOG:
+                if (mController != null) {
+                    mController.createScreenRecordDialog(this, null, null, null, null).show();
+                }
+                break;
         }
         return Service.START_STICKY;
     }
 
     @Override
     public IBinder onBind(Intent intent) {
-        return null;
+        return mBinder;
     }
 
     @Override
     public void onCreate() {
         super.onCreate();
+        mController.addCallback((RecordingController.RecordingStateChangeCallback) mBinder);
+    }
+
+    @Override
+    public void onDestroy() {
+        mController.removeCallback((RecordingController.RecordingStateChangeCallback) mBinder);
+        super.onDestroy();
     }
 
     @Nullable
@@ -755,4 +772,66 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
             stopService();
         }
     }
+
+    private class RecordingServiceBinder extends IRemoteRecording.Stub
+            implements RecordingController.RecordingStateChangeCallback {
+
+        private ArrayList<IRecordingCallback> mCallbackList = new ArrayList<>();
+
+        @Override
+        public void startRecording() throws RemoteException {
+            Intent intent = new Intent(RecordingService.this, RecordingService.class);
+            intent.setAction(ACTION_SHOW_DIALOG);
+            RecordingService.this.startService(intent);
+        }
+
+        @Override
+        public void stopRecording() throws RemoteException {
+            Intent intent = new Intent(RecordingService.this, RecordingService.class);
+            intent.setAction(ACTION_STOP_NOTIF);
+            RecordingService.this.startService(intent);
+        }
+
+        @Override
+        public boolean isRecording() throws RemoteException {
+            return mController.isRecording();
+        }
+
+        @Override
+        public boolean isStarting() throws RemoteException {
+            return mController.isStarting();
+        }
+
+        public void addRecordingCallback(IRecordingCallback callback) throws RemoteException {
+            if (!mCallbackList.contains(callback)) {
+                mCallbackList.add(callback);
+            }
+        }
+
+        public void removeRecordingCallback(IRecordingCallback callback) throws RemoteException {
+            mCallbackList.remove(callback);
+        }
+
+        @Override
+        public void onRecordingStart() {
+            for (IRecordingCallback callback : mCallbackList) {
+                try {
+                    callback.onRecordingStart();
+                } catch (RemoteException e) {
+                    // do nothing
+                }
+            }
+        }
+
+        @Override
+        public void onRecordingEnd() {
+            for (IRecordingCallback callback : mCallbackList) {
+                try {
+                    callback.onRecordingEnd();
+                } catch (RemoteException e) {
+                    // do nothing
+                }
+            }
+        }
+    }
 }
-- 
2.47.1

